### ISSUE TITLE
CDR stop event not emitted

### ISSUE THREAD

**Issue by spointer** [davehorton/drachtio-server#112]
Dear Dave.

When the UAC sends CANCEL SIP method, CDR stop event is not emitted.

I think drachtio-server sends the right 200 OK and 487(request terminated) but there is no CDR stop event and drachtio-srf raise the "Response not sent due to unknown transaction" exception.

Could you verify the following SIP logs?
Unfortunately, i can't provide the full drachtio-server logs because the UAS are deployed in the production mode.

SIP logs are as follows:

2020-02-18 09:01:56.413941 recv 962 bytes from udp/[62.255.184.19]:19999 at 09:01:56.413752:
INVITE sip:3308@172.25.0.2:3417 SIP/2.0
Via: SIP/2.0/UDP 10.1.32.14:19999;branch=z9hG4bK6bf4c3b062ab4089ad3cbb7334c84037
From: "3203"<sip:3203@10.1.32.14:19999>;tag=b252fa9cac434021be64f99671d606d0
To: <sip:3308@172.25.0.2:3417>
Call-ID: 2f3014f5c292441d96bb67d079988827@10.1.32.14:19999
CSeq: 2 INVITE
Max-Forwards: 70
Contact: <sip:3308@10.1.32.14:19999>
Allow: INVITE, ACK, CANCEL, OPTIONS, BYE, REFER, NOTIFY, MESSAGE, INFO
Supported: replaces
User-Agent: CALL-AGENT 2.1
Content-Type: application/sdp
Content-Length: 204

v=0
o=- 1582016516 1582016516 IN IP4 10.1.32.14
s=dialnet
c=IN IP4 10.1.32.14
t=0 0
m=audio 10962 RTP/AVP 101 8
a=fmtp:101 0-15
a=rtpmap:101 telephone-event/8000
a=rtpmap:8 PCMA/8000
a=sendrecv

2020-02-18 09:01:56.414196 send 345 bytes to udp/[62.255.184.19]:19999 at 09:01:56.414104:
SIP/2.0 100 Trying
Via: SIP/2.0/UDP 10.1.32.14:19999;branch=z9hG4bK6bf4c3b062ab4089ad3cbb7334c84037;received=62.255.184.19;rport=19999
From: "3203" <sip:3203@10.1.32.14:19999>;tag=b252fa9cac434021be64f99671d606d0
To: <sip:3308@172.25.0.2:3417>
Call-ID: 2f3014f5c292441d96bb67d079988827@10.1.32.14:19999
CSeq: 2 INVITE
Content-Length: 0

SIP/2.0 180 Ringing
Via: SIP/2.0/UDP 10.1.32.14:19999;branch=z9hG4bK6bf4c3b062ab4089ad3cbb7334c84037;received=62.255.184.19;rport=19999
From: "3203" <sip:3203@10.1.32.14:19999>;tag=b252fa9cac434021be64f99671d606d0
To: <sip:3308@172.25.0.2:3417>;tag=S30p72K798BgS
Call-ID: 2f3014f5c292441d96bb67d079988827@10.1.32.14:19999
CSeq: 2 INVITE
Content-Type: application/sdp
Content-Length: 239

v=0
o=- 1582016516 1582016518 IN IP4 131.120.5.124
s=cog
c=IN IP4 131.120.5.124
t=0 0
m=audio 10824 RTP/AVP 8 101
a=rtpmap:8 PCMA/8000
a=rtpmap:101 telephone-event/8000
a=fmtp:101 0-16
a=ptime:20
a=maxptime:150
a=sendrecv

2020-02-18 09:01:57.296360 recv 379 bytes from udp/[62.255.184.19]:19999 at 09:01:57.296240:
CANCEL sip:3308@172.25.0.2:3417 SIP/2.0
Via: SIP/2.0/UDP 10.1.32.14:19999;branch=z9hG4bK6bf4c3b062ab4089ad3cbb7334c84037
From: "3203"<sip:3203@10.1.32.14:19999>;tag=b252fa9cac434021be64f99671d606d0
To: <sip:3308@172.25.0.2:3417>
Call-ID: 2f3014f5c292441d96bb67d079988827@10.1.32.14:19999
CSeq: 2 CANCEL
Max-Forwards: 70
User-Agent: CALL-AGENT 2.1
Content-Length: 0

2020-02-18 09:01:57.296567 send 359 bytes to udp/[62.255.184.19]:19999 at 09:01:57.296478:
SIP/2.0 200 OK
Via: SIP/2.0/UDP 10.1.32.14:19999;branch=z9hG4bK6bf4c3b062ab4089ad3cbb7334c84037;received=62.255.184.19;rport=19999
From: "3203" <sip:3203@10.1.32.14:19999>;tag=b252fa9cac434021be64f99671d606d0
To: <sip:3308@172.25.0.2:3417>;tag=S30p72K798BgS
Call-ID: 2f3014f5c292441d96bb67d079988827@10.1.32.14:19999
CSeq: 2 CANCEL
Content-Length: 0

2020-02-18 09:01:57.296801 send 374 bytes to udp/[62.255.184.19]:19999 at 09:01:57.296729:
SIP/2.0 487 Request Terminated
Via: SIP/2.0/UDP 10.1.32.14:19999;branch=z9hG4bK6bf4c3b062ab4089ad3cbb7334c84037;received=62.255.184.19;rport=19999
From: "3203"<sip:3203@10.1.32.14:19999>;tag=b252fa9cac434021be64f99671d606d0
To: <sip:3308@172.25.0.2:3417>;tag=S30p72K798BgS
Call-ID: 2f3014f5c292441d96bb67d079988827@10.1.32.14:19999
CSeq: 2 INVITE
Content-Length: 0

2020-02-18 09:01:57.307629 recv 429 bytes from udp/[62.255.184.19]:19999 at 09:01:57.307532:
ACK sip:3308@172.25.0.2:3417 SIP/2.0
Via: SIP/2.0/UDP 10.1.32.14:19999;branch=z9hG4bK6bf4c3b062ab4089ad3cbb7334c84037
From: "3203"<sip:3203@10.1.32.14:19999>;tag=b252fa9cac434021be64f99671d606d0
To: <sip:3308@172.25.0.2:3417>;tag=S30p72K798BgS
Call-ID: 2f3014f5c292441d96bb67d079988827@10.1.32.14:19999
CSeq: 2 ACK
Contact: <sip:3308@10.1.32.14:19999>
User-Agent: CALL-AGENT 2.1
Max-Forwards: 70
Content-Length: 0

**Comment by spointer** [comment]
I found this closed issue and could handle the post processing when i did on being emitted cdr:stop event.
https://github.com/davehorton/drachtio-server/issues/29

But why not emitted cdr:stop event after 487 is sent to UAC?

**Comment by davehorton** [comment]
yes, this sounds like a bug.  I will check into it

**Comment by spointer** [comment]
Dear Dave,
I have a plan to separate drachtio UA application from CDR recording(i.e. creating CDR recording drachtio application).
If this issue is not resolved, the same cdr recording logic should be remained to UA application because as you know request cancel event is the UA application logic.

Any update on this issue?

**Comment by davehorton** [comment]
could you test a fix on the branch `uas-cancel-cdr-stop` ?

**Comment by spointer** [comment]
cdr stop event is emitted as expected after applying uas-cancel-cdr-stop branch(v0.8.6-rc1-8-gd25b958d).
"Response not sent due to unkown transaction" error is still raised on the drachtio-srf side but it seems this is not related to this issue.

Thank you.

**Comment by davehorton** [comment]
I'm unclear on the "Response not sent due to unkown transaction" error.  What are you doing on the drachtio-srf side that is leading to that?  There is no need to try to send a response when a request is canceled.

**Comment by spointer** [comment]
Ah, sorry for bothering you and please never mind.
I was mistaken in my application side. I'm using your drachtio-astmrf module and my application was trying to send internal INVITE(i.e, calling connectCaller function) to asterisk even after request cancel event emitted.

**Comment by spointer** [comment]
Anyway, i found the cdr stop event's msg object has no status code. 
I think it should have status and the value should be 487, isn't it?

**Comment by davehorton** [comment]
Yes, you are right.  I am sending the Cancel message, not the 487 response to the INVITE.  Let me look into that..
